### PR TITLE
A set of NCrypt and smartcard fixes

### PR DESCRIPTION
**Pull request by hardening**
This set fixes misc ncrypt and smartcard related things:
* fixing some invalid name constants
* fix build and search paths on MacOs
* add the reader property in ncrypt
* support the `\\.\<reader>\` syntax for `pszScope` in `NCryptEnumKeys()` which seems to be widely supported, even if not documented
* start a basic implementation of `SCardListCards` with a fixed list of ATR for now
* add a basic smartcard traffic parser

### CODE REVIEW

File: winpr/libwinpr/ncrypt/ncrypt.c [hunk 1]
```diff
@@ -126,9 +126,18 @@ SECURITY_STATUS NCryptOpenStorageProvider(NCRYPT_PROV_HANDLE* phProvider, LPCWST
...
+	if (_wcscmp(pszProviderName, MS_SMART_CARD_KEY_STORAGE_PROVIDER) == 0 ||
+		_wcscmp(pszProviderName, MS_SCARD_PROV) == 0)
 	{
-		static LPCSTR openscPaths[] = { "/usr/lib/x86_64-linux-gnu/pkcs11/opensc-pkcs11.so", NULL };
+		static LPCSTR openscPaths[] = {
+#ifdef __APPLE__
+				"/usr/local/lib/pkcs11/opensc-pkcs11.so",
+#else
+				/* linux and UNIXes */
+				"/usr/lib/x86_64-linux-gnu/pkcs11/opensc-pkcs11.so",
```
Comment (akallabeth) [inline winpr/libwinpr/ncrypt/ncrypt.c:137]: this looks wrong.
1. There are still 32bit builds out there
2. There are these that use the `lib64` style paths (e.g. fedora)

File: winpr/libwinpr/ncrypt/ncrypt_pkcs11.c [hunk 2]
```diff
@@ -447,11 +444,41 @@ static SECURITY_STATUS NCryptP11EnumKeys(NCRYPT_PROV_HANDLE hProvider, LPCWSTR p
...
+	if (pszScope)
+	{
+		/*
+		 * check whether pszScope is of the form \\.\<reader name>\ for filtering by
+		 * card reader
+		 */
+		char asciiScope[128 + 6];
+		int asciiScopeLen;
+
+		if (WideCharToMultiByte(CP_UTF8, 0, pszScope, _wcslen(pszScope) + 1, asciiScope,
```
Comment (akallabeth) [inline winpr/libwinpr/ncrypt/ncrypt_pkcs11.c:464]: Better use `sizeof(asciiScope)-1` to ensure '\0' termination

File: winpr/libwinpr/ncrypt/ncrypt_pkcs11.c [hunk 3]
```diff
@@ -583,15 +614,29 @@ static SECURITY_STATUS NCryptP11KeyGetProperties(NCryptP11KeyHandle* keyHandle,
...
 			break;
+		case NCRYPT_PROPERTY_READER: {
+			CK_SLOT_INFO slotInfo;
+
+			rv = provider->p11->C_GetSlotInfo(keyHandle->slotId, &slotInfo);
+			if (rv != CKR_OK)
+				return NTE_BAD_KEY;
+
+			fix_padded_string((char*)slotInfo.slotDescription, sizeof(slotInfo.slotDescription));
+			*pcbResult = 2 * (strlen((char*)slotInfo.slotDescription) + 1);
```
Comment (akallabeth) [inline winpr/libwinpr/ncrypt/ncrypt_pkcs11.c:629]: better use `strnlen` as the buffer is of fixed size

File: winpr/libwinpr/ncrypt/ncrypt_pkcs11.c [hunk 4]
```diff
@@ -447,11 +444,41 @@ static SECURITY_STATUS NCryptP11EnumKeys(NCRYPT_PROV_HANDLE hProvider, LPCWSTR p
...
+		int asciiScopeLen;
+
+		if (WideCharToMultiByte(CP_UTF8, 0, pszScope, _wcslen(pszScope) + 1, asciiScope,
+		                        sizeof(asciiScope), "?", NULL) <= 0)
+			return NTE_INVALID_PARAMETER;
+
+		if (strstr(asciiScope, "\\\\.\\") != asciiScope)
+			return NTE_INVALID_PARAMETER;
+
+		asciiScopeLen = strlen(asciiScope);
```
Comment (akallabeth) [inline winpr/libwinpr/ncrypt/ncrypt_pkcs11.c:471]: better use `strnlen` as the buffer is of fixed size

File: winpr/libwinpr/ncrypt/ncrypt_pkcs11.c [hunk 5]
```diff
@@ -401,7 +398,7 @@ static SECURITY_STATUS parseKeyName(LPCWSTR pszKeyName, CK_SLOT_ID* slotId, CK_B
 	char asciiKeyName[128];
```
Comment (akallabeth) [inline winpr/libwinpr/ncrypt/ncrypt_pkcs11.c:398]: better initialize variable with `={0};`

File: winpr/libwinpr/ncrypt/ncrypt_pkcs11.c [hunk 6]
```diff
@@ -583,15 +614,29 @@ static SECURITY_STATUS NCryptP11KeyGetProperties(NCryptP11KeyHandle* keyHandle,
...
+				return NTE_BAD_KEY;
+
+			fix_padded_string((char*)slotInfo.slotDescription, sizeof(slotInfo.slotDescription));
+			*pcbResult = 2 * (strlen((char*)slotInfo.slotDescription) + 1);
+			if (pbOutput)
+			{
+				if(cbOutput < *pcbResult)
+					return NTE_NO_MEMORY;
+
+				MultiByteToWideChar(CP_UTF8, 0, (LPCSTR)slotInfo.slotDescription, -1, (LPWSTR)pbOutput, cbOutput);
```
Comment (akallabeth) [inline winpr/libwinpr/ncrypt/ncrypt_pkcs11.c:635]: Missing return value check

File: winpr/libwinpr/ncrypt/test/TestNCryptSmartcard.c [hunk 7]
```diff
@@ -22,60 +22,128 @@
...
-		PBYTE certBytes = NULL;
+		NCRYPT_PROV_HANDLE provider;
+		char providerNameStr[256];
+		PVOID enumState = NULL;
+		size_t i = 0;
+		NCryptKeyName* keyName = NULL;
 
-		WideCharToMultiByte(CP_UTF8, 0, keyName->pszName, -1, keyNameStr, sizeof(keyNameStr), NULL,
-		                    FALSE);
+		WideCharToMultiByte(CP_UTF8, 0, names[j].pszName, -1, providerNameStr, sizeof(providerNameStr), NULL,
```
Comment (akallabeth) [inline winpr/libwinpr/ncrypt/test/TestNCryptSmartcard.c:83]: Missing return value check

File: winpr/libwinpr/ncrypt/test/TestNCryptSmartcard.c [hunk 8]
```diff
@@ -22,60 +22,128 @@
...
+			WCHAR reader[1024];
+			PBYTE certBytes = NULL;
 
-		certBytes = calloc(1, cbOutput);
-		status = NCryptGetProperty(phKey, NCRYPT_CERTIFICATE_PROPERTY, certBytes, cbOutput,
-		                           &cbOutput, dwFlags);
-		if (status != ERROR_SUCCESS)
-		{
-			WLog_ERR(TAG, "unable to retrieve certificate for key %s", keyNameStr);
+			WideCharToMultiByte(CP_UTF8, 0, keyName->pszName, -1, keyNameStr, sizeof(keyNameStr), NULL,
```
Comment (akallabeth) [inline winpr/libwinpr/ncrypt/test/TestNCryptSmartcard.c:100]: Missing return value check

File: winpr/libwinpr/ncrypt/test/TestNCryptSmartcard.c [hunk 9]
```diff
@@ -22,60 +22,128 @@
...
+			if (status != ERROR_SUCCESS)
+			{
+				WLog_ERR(TAG, "unable to open key %s", keyNameStr);
+				continue;
+			}
+
+			status = NCryptGetProperty(phKey, NCRYPT_READER_PROPERTY, (PBYTE)reader, sizeof(reader), &cbOutput, dwFlags);
+			if (status == ERROR_SUCCESS)
+			{
+				char readerStr[1024];
```
Comment (akallabeth) [inline winpr/libwinpr/ncrypt/test/TestNCryptSmartcard.c:115]: Initialize and ensure '\0' termination afterward

File: winpr/libwinpr/ncrypt/ncrypt_pkcs11.c [hunk 10]
```diff
@@ -583,15 +614,29 @@ static SECURITY_STATUS NCryptP11KeyGetProperties(NCryptP11KeyHandle* keyHandle,
...
-	/* TODO: shall adjust objectFilter and objectFilterLen depending on requested
-	 * the property when we add new ones */
 	switch (property)
 	{
 		case NCRYPT_PROPERTY_CERTIFICATE:
 			break;
+		case NCRYPT_PROPERTY_READER: {
+			CK_SLOT_INFO slotInfo;
+
+			rv = provider->p11->C_GetSlotInfo(keyHandle->slotId, &slotInfo);
```
Comment (akallabeth) [inline winpr/libwinpr/ncrypt/ncrypt_pkcs11.c:624]: Better assert pointer dereferences here to make tracing issues easier

File: winpr/libwinpr/ncrypt/test/TestNCryptSmartcard.c [hunk 11]
```diff
@@ -22,60 +22,128 @@
...
-	while ((status = NCryptEnumKeys(provider, NULL, &keyName, &enumState, NCRYPT_SILENT_FLAG)) ==
-	       ERROR_SUCCESS)
+	for (j = 0; j < providerCount; j++)
 	{
-		NCRYPT_KEY_HANDLE phKey;
-		DWORD dwFlags = 0, cbOutput;
-		char keyNameStr[256];
-		PBYTE certBytes = NULL;
+		NCRYPT_PROV_HANDLE provider;
+		char providerNameStr[256];
```
Comment (akallabeth) [inline winpr/libwinpr/ncrypt/test/TestNCryptSmartcard.c:78]: Initialize and ensure '\0' termination afterward

File: winpr/libwinpr/ncrypt/test/TestNCryptSmartcard.c [hunk 12]
```diff
@@ -22,60 +22,128 @@
...
+		while ((status = NCryptEnumKeys(provider, NULL, &keyName, &enumState, NCRYPT_SILENT_FLAG)) ==
+			   ERROR_SUCCESS)
 		{
-			WLog_ERR(TAG, "unable to retrieve certificate for key %s", keyNameStr);
-			NCryptFreeObject((NCRYPT_HANDLE)phKey);
-			continue;
-		}
+			NCRYPT_KEY_HANDLE phKey;
+			DWORD dwFlags = 0, cbOutput;
+			char keyNameStr[256];
```
Comment (akallabeth) [inline winpr/libwinpr/ncrypt/test/TestNCryptSmartcard.c:96]: Initialize and ensure '\0' termination afterward

**Review by akallabeth — CHANGES_REQUESTED** [review]
Overall LGTM.

Some more checks might be a nice to have thing to avoid problems tracing wrong parameters/incomplete struct initialization

File: winpr/libwinpr/ncrypt/ncrypt.c (hunk 1 above)
Comment (hardening) [reply winpr/libwinpr/ncrypt/ncrypt.c]: Partially fixed (some paths added should cover most cases).

File: winpr/libwinpr/ncrypt/ncrypt_pkcs11.c [hunk 13]
```diff
@@ -727,11 +772,12 @@ SECURITY_STATUS NCryptOpenP11StorageProviderEx(NCRYPT_PROV_HANDLE* phProvider,
 
 	while (*modulePaths)
```
[the remark this answers is not in the document]
Comment (hardening) [reply winpr/libwinpr/ncrypt/ncrypt_pkcs11.c:775]: Done in the caller.

File: winpr/libwinpr/ncrypt/ncrypt_pkcs11.c (hunk 10 above)
Comment (hardening) [reply winpr/libwinpr/ncrypt/ncrypt_pkcs11.c:624]: I don't see what is more likely to happen here that would require an assert. And BTW segfaulting here would make a clean and nice segfault easily debuggable.

**Review by akallabeth — APPROVED** [review]
LGTM

**Pull request merged**